Mathlib/Combinatorics/SetFamily/Lym.lean

```diff
@@ -0,0 +1,260 @@
+/-
+Copyright (c) 2022 Bhavik Mehta, Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Alena Gusakov, Yaël Dillies
+
+! This file was ported from Lean 3 source module combinatorics.set_family.lym
+! leanprover-community/mathlib commit 861a26926586cd46ff80264d121cdb6fa0e35cc1
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.BigOperators.Ring
+import Mathlib.Algebra.Order.Field.Basic
+import Mathlib.Combinatorics.DoubleCounting
+import Mathlib.Combinatorics.SetFamily.Shadow
+import Mathlib.Data.Rat.Order
+
+/-!
+# Lubell-Yamamoto-Meshalkin inequality and Sperner's theorem
+
+This file proves the local LYM and LYM inequalities as well as Sperner's theorem.
+
+## Main declarations
+
+* `Finset.card_div_choose_le_card_shadow_div_choose`: Local Lubell-Yamamoto-Meshalkin inequality.
+  The shadow of a set `𝒜` in a layer takes a greater proportion of its layer than `𝒜` does.
+* `Finset.sum_card_slice_div_choose_le_one`: Lubell-Yamamoto-Meshalkin inequality. The sum of
+  densities of `𝒜` in each layer is at most `1` for any antichain `𝒜`.
+* `IsAntichain.sperner`: Sperner's theorem. The size of any antichain in `Finset α` is at most the
+  size of the maximal layer of `Finset α`. It is a corollary of `sum_card_slice_div_choose_le_one`.
+
+## TODO
+
+Prove upward local LYM.
+
+Provide equality cases. Local LYM gives that the equality case of LYM and Sperner is precisely when
+`𝒜` is a middle layer.
+
+`falling` could be useful more generally in grade orders.
+
+## References
+
+* http://b-mehta.github.io/maths-notes/iii/mich/combinatorics.pdf
+* http://discretemath.imp.fu-berlin.de/DMII-2015-16/kruskal.pdf
+
+## Tags
+
+shadow, lym, slice, sperner, antichain
+-/
+
+
+open Finset Nat
+
+open BigOperators FinsetFamily
+
+variable {𝕜 α : Type _} [LinearOrderedField 𝕜]
+
+namespace Finset
+
+/-! ### Local LYM inequality -/
+
+
+section LocalLym
+ -- porting note: added `Decidable` for ⊆
```

Comment:
not sure how to handle the fact that I couldn't do `card_mul_le_card_shadow_mul` without the Decidable assumption for `⊆`. Is there some way around it?
Should this be placed elsewhere?